### PR TITLE
Fix amortized asymptotics of `BytesMut`

### LINKED ISSUE #524: Bad “optimization” of `BytesMut` re-allocations
Reading 

https://github.com/tokio-rs/bytes/blob/68afb40df9ea2e61a93d0e26f6f8a9f231979452/src/bytes_mut.rs#L577-L579

I felt like this can completely kill the asymptotics of `BytesMut`. And sure enough, e.g. this example:

on a full `BytesMut`, a million bytes, doing 10'000 `get_u8` and `put_u8` operations moves the entire contents of the `BytesMut` 10'000 times, resulting in a run-time (in the playground) of up to 1 second, instead of less than a millisecond.

```rs
use bytes::{BytesMut, BufMut, Buf};

use std::time;

fn main() {
    let mut bytes = BytesMut::with_capacity(1024);
    bytes.put_bytes(0, 1_000_000);
    assert_eq!(bytes.capacity(), 1_000_000);
    let start = time::Instant::now();
    for _ in 0..10_000 {
        bytes.get_u8();
        bytes.put_u8(0);
    }
    println!("{:>9.3?} - 10'000 × get+put (each moving all the data)", start.elapsed());
    let start = time::Instant::now();
    bytes.put_u8(0);
    println!("{:>9.3?} - one put (re-allocating)", start.elapsed());
    let start = time::Instant::now();
    for _ in 0..10_000 {
        bytes.get_u8();
        bytes.put_u8(0);
    }
    println!("{:>9.3?} - 10'000 × get+put (no allocations or moves)", start.elapsed());
}
```
```plain
633.147ms - 10'000 × get+put (each moving all the data)
 71.792µs - one put (re-allocating)
223.719µs - 10'000 × get+put (no allocations or moves)
```
([playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=61594b1797a3faab78e715101463f29b))

The optimization here is bad, since usually automatic re-allocations follow the principle that the capacity has to grow by a constant factor each time, whereas moving the data might only increase the capacity by a tiny bit. This isn't a re-allocation; it just copies all the data, but really the actual “allocation” part of a re-allocation is cheap, it always is the copying that it's all about.

Vec (currently) uses a [factor 2](https://doc.rust-lang.org/1.57.0/src/alloc/raw_vec.rs.html#440), so I suggest that this “optimization” should really only ever trigger if the gained capacity, i.e. `self.capacity() - self.len() + off`, increases the previous capacity, `self.capacity()` at least by a factor of two, i.e. if
* (current condition still relevant): `self.capacity() - self.len() + off >= additional`, **and also**
* (new condition): `off >= self.capacity()`

Interesting consequence: This also has the effect that `copy_nonoverlapping` can be used.

### PR DESCRIPTION
**Pull request by steffahn**
Closes #524. See that issue for a description of what this PR changes and why.

### CODE REVIEW

**Comment by steffahn** [comment]
Interesting test case that fails here; there's room for improvement so this doesn't have to fail.

**Comment by steffahn** [comment]
So, I somehow only considered the case of reserving more space when the buffer is currently full to capacity. I suppose that it's not actually an important metric that the capacity is doubled when reserving new size, but rather it's sufficient to consider the current length. After all, the _cost_ of moving the data is only as high as the current length.

Let me try to remember the amortized analysis (for `Vec<u8>`) for why pushing bytes one-by-one is (amortized) constant time (for each `push` operation)... the re-allocating+moving a whole, full, `Vec` is accounted for by previous push operations. If the capacity of a `Vec` doubles every time, than always the cost of (the latter) half of the data that had been pushed to the `Vec` accounts for moving the whole thing another time. Each byte being pushed allows for two bytes being copied in the future, thus we get constant time push. After resizing, the latter half of the `Vec` is empty and would need to be filled by push operations again, before more resizing could happen.

The test case in question uses `reserve` though, requiring additional capacity without a previously completely filled buffer, which complicates the matter because I'm not 100% certain how that interacts with the whole analysis.

Well, at least, one main point in the analysis above was that when we *know* that _some amount_ of amortized-constant-time operations have happened since the last time all the data was (possibly re-allocated and) copied, it ought to be okay to copy _twice_ the amount of data around, compared to how many of those amortized-constant-time operations operations we had, as long as we and up with an at-most half-full buffer in the end. (The last condition to make sure that ordinary sequences of `push` will write half the buffer before resizing / copying again.)

So in this test case
```rs
fn bytes_buf_mut_reuse_when_fully_consumed() {
    use bytes::{Buf, BytesMut};
    let mut buf = BytesMut::new();
    buf.reserve(8192);
    buf.extend_from_slice(&[0u8; 100][..]);

    let p = &buf[0] as *const u8;
    buf.advance(100);

    buf.reserve(8192);
    buf.extend_from_slice(b" ");

    assert_eq!(&buf[0] as *const u8, p);
}
```
there's, for the second `reserve` (which is the operation in question), an empty buffer with an offset of 100. We know from an offset of 100, that 100 bytes have been _read_ (using `get_u8` or similar) since the last time the data was moved to the front of the `Vec`. Reading a byte in this manner is an amortized-constant time (well, even an _actual_ constant time) operation, so we can use that as justification to move up to `200` bytes, as long as we end up with a `capacity >= 2*len` situation in the end.

The `2` is actually an arbitrary constant factor, coming from the arbitrary rule that "each constant-time operation accounts for two bytes being copied in the future" let's generalize to `C`

This suggests a rule like...
* `C * off >= len` should be checked (this is weaker than the current [in this PR!] condition `off >= cap`, provided `C >= 1`) to make sure we can afford moving all the data
* `2 * len <= off + cap` to make sure that the buffer is half-empty after the operation (`off + cap` is the new capacity), so that future resizes due to actually reaching the limit can be afforded. This is also weaker than the current [in this PR!] condition `off >= cap`. Not quite as obvious, but: if `off >= cap`, then `2 * len <= 2 * cap == cap + cap <= off + cap`

Also, I like keeping the `copy_nonoverlapping`, and the test case that failed doesn't need to break it. Let's see how the calculation here works out then:

* Overlap is avoided if `off >= len`.
* This implies that `C == 1` is a great choice, actually!

<hr>

So with `C == 1`, the checks are

* `off >= len`
* `2 * len <= off + cap`

except, the second one is redundant now. `len + len <= off + cap` is clear from `len <= off` and `len <= cap`.

So let me change the PR to just use a `off >= len` check instead of `off >= cap` check.

**Comment by steffahn** [comment]
I just noticed the part of [the documentation](https://docs.rs/bytes/1.1.0/bytes/struct.BytesMut.html#method.reserve) that says

> Before allocating new buffer space, the function will attempt to reclaim space in the existing buffer. If the current handle references a small view in the original buffer and all other handles have been dropped, and the requested capacity is less than or equal to the existing buffer’s capacity, then the current view will be copied to the front of the buffer and the handle will take ownership of the full buffer.

Is shockingly explicit about the “the requested capacity is less than or equal to the existing buffer’s capacity” condition. I’m expecting a more vague statement similar to the preceding “More than additional bytes may be reserved in order to avoid frequent reallocations.” in order to leave the implementation details unspecified. As it stands, this sounds as if only frequent re**allocations** are bad, but frequent "copy everything to the front of the buffer" is totally okay.

<hr>

I’m open for suggestions of how to best improve this.

**Comment by Darksonn** [comment]
As long as we have `C * len ≤ cap` for some constant `C` greater than one, then the amortization will work since you have to perform at least `(C-1) * len` get+put operations between each time it copies the data backwards.

**Comment by steffahn** [comment]
Interesting point. So checking `2 * len <= off + cap` alone could be enough?

**Comment by Darksonn** [comment]
Yes, though arguably we might want to pick a smaller `C` such as 1.5.

**Comment by steffahn** [comment]
I'm trying to evaluate whether it's important to make sure that something like
```rs
let mut bytes = BytesMut::with_capacity(10_000);
bytes.put_bytes(0, 5_000);
loop {
    bytes.get_u8();
    bytes.put_u8(0);
    bytes.reserve(5_000);
}
```
remains efficient.

The buffer is always half-full at the point of `reserve`.

An approach that allows shifting the data when `2 * len <= off + cap` will shift the whole (`5_000` bytes) data for each `reserve`, while the PR as it stands currently would increase the size of the buffer once by growing the `Vec` (so it goes to `20_000`), and then it only needs to shift all the `5_000` bytes once every `20_000` iterations.

**Comment by Darksonn** [comment]
> increase the size of the buffer once by growing the Vec (so it goes to 20_000), and then it only needs to shift all the 5_000 bytes once every 20_000 iterations.

That seems fine to me.

**Comment by Darksonn** [comment]
What's the status here? If you wanted to update the text in the documentation, then go ahead and do that as well.

**Comment by steffahn** [comment]
Oh, well I guess I kind-of forgot about this PR entirely. Alright, so I’m going to think of a way to reword the documentation, and I’ll keep the implementation as it currently is in this PR.

**Comment by Darksonn** [comment]
Sounds good!

**Comment by NobodyXu** [comment]
@steffahn I think you forgot to also fix the amortized asumptotics for cases where `BytesMut` holds a unique reference to `KIND_ARC`.

**Comment by steffahn** [comment]
@NobodyXu I will take a look at this this weekend.

**Comment by steffahn** [comment]
Sorry for the delay. The new documentation got somewhat lengthy, but should be okay now. I will probably also re-read it for any mistakes tomorrow.

@NobodyXu, I indeed didn’t look at the `KIND_ARC` case before, and I think it makes sense to apply the same logic there. I have adapted the same logic into that part of the code now (hopefully correctly).

File: src/bytes_mut.rs [hunk 1]
```diff
@@ -511,11 +511,24 @@ impl BytesMut {
...
+    /// 
+    /// There are additional constraints for when this reclaiming can happen,
+    /// in order to prevent access patterns with surprising amounts of expensive
+    /// copying operations. If there is nothing to copy (i.e. the current view
+    /// is empty), then the reclaiming _always_ happens whenever the original
+    /// buffer is sufficiently large. Otherwise, as of this writing, the buffer
+    /// will -- even if it is sufficiently large -- only be reclaimed if the
+    /// offset of the current view (from the start of the original buffer) is be
+    /// greater than or equal to the length of the current view, but this
+    /// precise condition might change in the future.
```
Comment (Darksonn) [inline src/bytes_mut.rs:531]: I think this could be improved and shortened quite a bit.

Comment (steffahn) [reply src/bytes_mut.rs]: Making stuff concise is the hard part, but I guess can try xD.

Not today anymore though.

**Comment by steffahn** [comment]
Fixed the trailing whitespace. I’ll have to figure out what the most proper way of calculating the offset without `offset_from` is. I could convert into `usize` and subtract; I’m not sure whether or not maybe the `BytesMut` type itself already stores the offset somehow. I won’t look back into this before tomorros.

**Comment by NobodyXu** [comment]
> Fixed the trailing whitespace. I’ll have to figure out what the most proper way of calculating the offset without `offset_from` is. I could convert into `usize` and subtract; I’m not sure whether or not maybe the `BytesMut` type itself already stores the offset somehow. I won’t look back into this before tomorrow.

I think you can substract `vec.capacity() - self.cap` and that would give you the offset.

Converting to `usize` is probably UB because of pointer provenance.

**Comment by Darksonn** [comment]
Converting into usize is definitely fine. The trouble is when you convert the usize _back_ into a pointer.

**Comment by NobodyXu** [comment]
> Converting into usize is definitely fine. The trouble is when you convert the usize _back_ into a pointer.

Yes, I think converting it back to a pointer might trigger UB, or at least involves some dark corner of the compiler.

**Comment by NobodyXu** [comment]
This PR seems to be staling for a long time...Will it be a good idea for me to fork @steffahn 's branch and create a separate PR as I want this optimization to be in bytes @Darksonn

**Comment by Darksonn** [comment]
I haven't looked at this for a while and I don't remember what the status is.

**Comment by steffahn** [comment]
IIRC, I only have to rephrase/improve the documentation. Sorry for the delay; in the meantime, I had to finish a master’s thesis due to the end of March; since that’s done now, I suppose I should get around to finish up this PR.

**Comment by Darksonn** [comment]
Sounds good. And congratulations on finishing your master's thesis!

**Comment by steffahn** [comment]
So, I’ve rewritten the section in the docs now. Maybe still not the perfect balance in terms of lengthiness, but it seems necessary to give at least enough information for the subsequent examples to make sense, and the condition is somewhat complex. I’m of course still open to suggestions how to make it shorter and/or better; but maybe it's already acceptable for now?

**Comment by Darksonn** [comment]
Not sure why, but it seems like some tests have started failing.

File: src/bytes_mut.rs [hunk 2]
```diff
@@ -625,7 +652,7 @@ impl BytesMut {
         let original_capacity_repr;
 
         unsafe {
-            original_capacity_repr = (*shared).original_capacity_repr;
+            original_capacity_repr = (*shared).original_capacity_repr;
```
Comment (Darksonn) [inline src/bytes_mut.rs:655]: This is probably causing a rustfmt CI failure.

**Review by Darksonn — APPROVED** [review]

File: src/bytes_mut.rs (hunk 2 above)
Comment (steffahn) [reply src/bytes_mut.rs:655]: Oh, that's an oversight / accidental whitespace modification.

Comment (steffahn) [reply src/bytes_mut.rs]: I squashed the two commits from today and fixed this whitespace.

**Comment by steffahn** [comment]
I can reproduce the test failure and miri UB locally when merging with master, I will need to take a look into this.

**Comment by Darksonn** [comment]
It may be due to some changes in miri? Note that #542 exists, which may affect things when it comes to miri.

**Comment by Darksonn** [comment]
Though it seems like the failure also exists on non-miri.

**Comment by steffahn** [comment]
There's the rest failure unrelated to miri though. And the miri error is reading of uninitialized memory. Tests and miri do succeed (with current miri) both in master and on this PR branch, but not on the result of merging the two. Maybe the diff just results in logically broken code somehow due to other changes (yet its not too broken to be a merge conflict), who knows... I hope I get around to looking into this tomorrow.

File: src/bytes_mut.rs [hunk 3]
```diff
@@ -636,11 +663,18 @@ impl BytesMut {
                 // sure that the vector has enough capacity.
                 let v = &mut (*shared).vec;
 
-                if v.capacity() >= new_cap {
-                    // The capacity is sufficient, reclaim the buffer
-                    let ptr = v.as_mut_ptr();
+                let v_capacity = v.capacity();
+                let ptr = v.as_mut_ptr();
+                let offset = self.ptr.as_ptr().offset_from(ptr) as usize;
```
Comment (NobodyXu) [inline src/bytes_mut.rs:668]: You could use:

```suggestion
                let offset = v_capacity - self.cap;
```

to calculate the offset here.

**Comment by NobodyXu** [comment]
Pinging this PR.

**Comment by NobodyXu** [comment]
@steffahn Pinging this PR.

**Comment by NobodyXu** [comment]
@steffahn Pinging since I'd really want this PR to be merged.

**Pull request closed without merging**